File: altcryp.php
<?php
// ============================================================
// ALTCRYP AUTO BOT - PHP (Cookie Only) - FIX V6
// Stop otomatis setelah 2x claim gagal berturut-turut
// ============================================================

error_reporting(E_ALL);
date_default_timezone_set('Asia/Jakarta');

define('MERAH', "\033[0;31m");
define('HIJAU', "\033[0;32m");
define('KUNING', "\033[0;33m");
define('CYAN', "\033[0;36m");
define('PUTIH', "\033[0;37m");
define('RESET', "\033[0m");
define('BOLD', "\033[1m");
define('DIM', "\033[2m");

define('BASE_URL', 'https://altcryp.com');
define('SOLVER_BASE', 'https://bypassallshortlinks.space');
define('CONFIG_FILE', 'config_altcryp.json');
define('CLAIM_INTERVAL', 300);  // default 5 menit
define('RETRY_INTERVAL', 60);
define('MAX_FAILURES', 2);      // <-- batas gagal sebelum stop

class CookieJar {
    private $cookies = [];
    
    public function set($name, $value) {
        $this->cookies[$name] = $value;
    }
    
    public function get($name) {
        return isset($this->cookies[$name]) ? $this->cookies[$name] : null;
    }
    
    public function getAll() {
        return $this->cookies;
    }
    
    public function parseSetCookie($header) {
        preg_match_all('/Set-Cookie:\s*([^;]+)/i', $header, $matches);
        if (!empty($matches[1])) {
            foreach ($matches[1] as $cookie_line) {
                $parts = explode('=', $cookie_line, 2);
                if (count($parts) == 2) {
                    $name = trim($parts[0]);
                    $value = trim($parts[1]);
                    if ($value === 'deleted' || strpos($cookie_line, 'expires=') !== false && strtotime(explode('expires=', $cookie_line)[1]) < time()) {
                        unset($this->cookies[$name]);
                    } else {
                        $this->cookies[$name] = $value;
                    }
                }
            }
        }
    }
    
    public function toString() {
        $str = '';
        foreach ($this->cookies as $k => $v) {
            $str .= "$k=$v; ";
        }
        return rtrim($str, '; ');
    }
    
    public function fromString($str) {
        if (empty($str)) return;
        $pairs = explode(';', $str);
        foreach ($pairs as $pair) {
            $parts = explode('=', trim($pair), 2);
            if (count($parts) == 2) {
                $this->cookies[$parts[0]] = $parts[1];
            }
        }
    }
}

function clear() { system('clear'); }

function timer($seconds, $prefix = "[!] Tunggu") {
    $wait = (int)$seconds;
    $frames = ['⣾', '⣽', '⣻', '⢿', '⡿', '⣟', '⣯', '⣷'];
    $current = 0;
    while ($wait > 0) {
        $start = microtime(true);
        while ((microtime(true) - $start) < 1) {
            $hours = floor($wait / 3600);
            $minutes = floor(($wait % 3600) / 60);
            $secs = $wait % 60;
            $time_str = sprintf('%02d:%02d:%02d', $hours, $minutes, $secs);
            $spinner = $frames[$current];
            echo PUTIH . $prefix . HIJAU . " $time_str " . PUTIH . $spinner . "\r";
            usleep(100000);
            $current = ($current + 1) % count($frames);
            if ((microtime(true) - $start) >= 1) break;
        }
        $wait--;
    }
    echo str_repeat(" ", 50) . "\r";
}

function get_config() {
    return file_exists(CONFIG_FILE) ? json_decode(file_get_contents(CONFIG_FILE), true) : null;
}

function save_config($config) {
    file_put_contents(CONFIG_FILE, json_encode($config, JSON_PRETTY_PRINT));
}

function get_cookie_from_user() {
    echo PUTIH . "Cookie (copy dari browser, format: key1=value1; key2=value2): " . KUNING;
    $cookie = trim(fgets(STDIN));
    return $cookie;
}

function get_apikey_from_user() {
    echo PUTIH . "API Key bypassallshortlinks: " . KUNING;
    $apikey = trim(fgets(STDIN));
    return $apikey;
}

function http_request($url, $method = 'GET', $data = [], $headers = [], CookieJar &$jar = null) {
    if ($jar === null) {
        $jar = new CookieJar();
    }
    
    $ch = curl_init();
    $options = [
        CURLOPT_URL => $url,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HEADER => true,
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_MAXREDIRS => 10,
        CURLOPT_SSL_VERIFYPEER => false,
        CURLOPT_SSL_VERIFYHOST => false,
        CURLOPT_TIMEOUT => 30,
        CURLOPT_CONNECTTIMEOUT => 30,
        CURLOPT_USERAGENT => 'Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Mobile Safari/537.36',
        CURLOPT_HTTPHEADER => [
            'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            'Accept-Language: id-ID,en;q=0.9',
            'Cache-Control: max-age=0',
            'Upgrade-Insecure-Requests: 1',
        ],
    ];
    
    $cookie_str = $jar->toString();
    if (!empty($cookie_str)) {
        $options[CURLOPT_COOKIE] = $cookie_str;
    }
    
    if (strtoupper($method) === 'POST') {
        $options[CURLOPT_POST] = true;
        $options[CURLOPT_POSTFIELDS] = http_build_query($data);
    }
    
    if (!empty($headers)) {
        $options[CURLOPT_HTTPHEADER] = array_merge($options[CURLOPT_HTTPHEADER], $headers);
    }
    
    curl_setopt_array($ch, $options);
    $response = curl_exec($ch);
    
    if ($response === false) {
        echo MERAH . "[!] cURL error: " . curl_error($ch) . RESET . "\n";
        return null;
    }
    
    $header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
    $body = substr($response, $header_size);
    $header = substr($response, 0, $header_size);
    
    $jar->parseSetCookie($header);
    
    return $body;
}

function solve_turnstile($sitekey, $pageurl, $apikey, $show_progress = false) {
    $url = SOLVER_BASE . "/in.php?key=" . urlencode($apikey) . 
           "&method=turnstile&sitekey=" . urlencode($sitekey) . 
           "&pageurl=" . urlencode($pageurl);
    
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0');
    $result = curl_exec($ch);
    if (!$result || !str_starts_with($result, 'OK|')) {
        echo MERAH . "[!] Gagal submit Turnstile: $result" . RESET . "\n";
        return null;
    }
    
    $task_id = explode('|', $result)[1];
    if ($show_progress) {
        echo KUNING . "[+] Task ID    : $task_id" . RESET . "\n";
        echo CYAN . "[*] Menunggu Turnstile..." . RESET . "\n";
    }
    
    for ($i = 0; $i < 45; $i++) {
        sleep(2);
        $poll_url = SOLVER_BASE . "/res.php?id=" . urlencode($task_id) . "&key=" . urlencode($apikey);
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $poll_url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        $result = curl_exec($ch);
        if ($result && str_starts_with($result, 'OK|')) {
            $token = explode('|', $result)[1];
            if ($show_progress) {
                echo HIJAU . "[+] Turnstile  : SOLVED ✓" . RESET . "\n";
            }
            return $token;
        }
        if ($show_progress && $i % 5 == 0) {
            echo DIM . "[*] Polling ke- " . ($i+1) . "/45..." . RESET . "\n";
        }
    }
    echo MERAH . "[!] Turnstile timeout." . RESET . "\n";
    return null;
}

function is_logged_in($jar) {
    $urls = [
        BASE_URL . '/dashboard',
        BASE_URL . '/faucet/currency/pol',
    ];
    
    foreach ($urls as $url) {
        $resp = http_request($url, 'GET', [], [], $jar);
        if (!$resp) continue;
        
        $debug_file = 'debug_' . md5($url) . '.html';
        file_put_contents($debug_file, $resp);
        
        if (strpos($resp, 'Logout') !== false || strpos($resp, 'Account') !== false || strpos($resp, 'Withdraw') !== false) {
            return true;
        }
        if (strpos($resp, 'csrf_token_name') !== false && strpos($resp, 'token') !== false) {
            return true;
        }
        if (strlen($resp) < 1000 && (strpos($resp, 'login') !== false || strpos($resp, 'Sign in') !== false)) {
            echo MERAH . "[!] Detected login page from $url" . RESET . "\n";
            return false;
        }
    }
    return false;
}

function get_coins($jar) {
    $html = http_request(BASE_URL, 'GET', [], [], $jar);
    if (!$html) return [];
    preg_match_all('/\/faucet\/currency\/([a-z]+)/', $html, $matches);
    if (!empty($matches[1])) {
        return array_values(array_unique($matches[1]));
    }
    return [];
}

function get_csrf_and_token($jar, $coin) {
    $url = BASE_URL . "/faucet/currency/$coin";
    $headers = ['Referer: ' . BASE_URL];
    $html = http_request($url, 'GET', [], $headers, $jar);
    if (!$html) return ['status' => 'error', 'msg' => 'No response'];
    
    file_put_contents('debug_faucet_' . $coin . '.html', $html);
    
    if (strpos($html, 'Login') !== false && strlen($html) < 500) {
        return ['status' => 'expired'];
    }
    if (strpos($html, 'csrf_token_name') === false || strpos($html, 'token') === false) {
        echo KUNING . "[!] CSRF/Token tidak ditemukan. Cuplikan HTML:\n" . RESET;
        echo substr($html, 0, 500) . "...\n";
        return ['status' => 'error', 'msg' => 'Missing CSRF/Token'];
    }
    
    // ===== REGEX YANG DIPERBAIKI =====
    if (preg_match('/name="csrf_token_name"[^>]*value="([^"]+)"/', $html, $m)) {
        $csrf = $m[1];
    } else {
        if (preg_match('/csrf_token_name"\s*value="([^"]+)"/', $html, $m)) {
            $csrf = $m[1];
        }
    }
    
    if (preg_match('/name="token"[^>]*value="([^"]+)"/', $html, $m)) {
        $token = $m[1];
    } else {
        if (preg_match('/token"\s*value="([^"]+)"/', $html, $m)) {
            $token = $m[1];
        }
    }
    
    if (empty($csrf) || empty($token)) {
        echo KUNING . "[!] CSRF atau Token kosong. Cek debug_faucet_$coin.html" . RESET . "\n";
        return ['status' => 'error', 'msg' => 'Missing CSRF/Token'];
    }
    
    return ['status' => 'ok', 'csrf' => $csrf, 'token' => $token];
}

function claim_faucet($jar, $coin, $apikey) {
    echo CYAN . "[*] Claim      : PROCESSING... (coin: " . strtoupper($coin) . ")" . RESET . "\n";
    
    $data = get_csrf_and_token($jar, $coin);
    if ($data['status'] === 'expired') {
        echo MERAH . "[!] Session expired atau cookie ga valid." . RESET . "\n";
        return 'EXPIRED';
    }
    if ($data['status'] !== 'ok') {
        echo MERAH . "[!] Gagal ambil CSRF/Token: " . $data['msg'] . RESET . "\n";
        return false;
    }
    
    $csrf = $data['csrf'];
    $token = $data['token'];
    echo KUNING . "[+] CSRF Token : $csrf" . RESET . "\n";
    echo KUNING . "[+] Token      : $token" . RESET . "\n";
    
    $sitekey = "0x4AAAAAAAHPLPJjjJUpAitl";
    $pageurl = BASE_URL . "/faucet/currency/$coin";
    $turnstile = solve_turnstile($sitekey, $pageurl, $apikey, true);
    if (!$turnstile) return false;
    
    $post_data = [
        'username_fake_field' => '',
        'csrf_token_name' => $csrf,
        'token' => $token,
        'captcha' => 'turnstile',
        'cf-turnstile-response' => $turnstile
    ];
    
    $url_verify = BASE_URL . "/faucet/verify/$coin";
    $result = http_request($url_verify, 'POST', $post_data, [
        'Content-Type: application/x-www-form-urlencoded',
        'Origin: ' . BASE_URL,
        'Referer: ' . BASE_URL . "/faucet/currency/$coin",
    ], $jar);
    
    if (!$result) {
        echo MERAH . "[!] Claim gagal (ga ada respon)." . RESET . "\n";
        return false;
    }
    
    if (strpos($result, 'login') !== false && strlen($result) < 500) {
        echo MERAH . "[!] Session expired setelah claim." . RESET . "\n";
        return 'EXPIRED';
    }
    
    if (strpos($result, 'Good job') !== false || strpos($result, 'success') !== false) {
        echo HIJAU . "[+] Claim      : SUCCESS ✓" . RESET . "\n";
        if (preg_match('/([\d.]+)\s+(\w+)/', $result, $reward)) {
            echo KUNING . "[+] Reward     : " . $reward[1] . " " . $reward[2] . RESET . "\n";
        }
        return true;
    }
    
    if (strpos($result, 'already claimed') !== false || strpos($result, 'wait') !== false) {
        echo KUNING . "[!] Claim      : COOLDOWN ⏳" . RESET . "\n";
        return false;
    }
    
    echo MERAH . "[?] Claim ga jelas." . RESET . "\n";
    file_put_contents('claim_debug_altcryp.html', $result);
    return false;
}

// ========== MAIN ==========
clear();

echo CYAN . "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" . RESET . "\n";
echo BOLD . KUNING . "              🚀 ALTCRYP AUTO BOT (PHP) - FIX V6" . RESET . "\n";
echo CYAN . "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" . RESET . "\n";
echo KUNING . "[!] Bot akan berhenti otomatis setelah 2x gagal berturut-turut." . RESET . "\n";

$config = get_config();
if (!$config) {
    echo KUNING . "[*] Config ga ketemu. Bikin baru..." . RESET . "\n";
    $cookie = get_cookie_from_user();
    $apikey = get_apikey_from_user();
    $config = ['cookie' => $cookie, 'apikey' => $apikey];
    save_config($config);
    echo HIJAU . "[+] Config disimpan di " . CONFIG_FILE . RESET . "\n";
    sleep(1);
}

$cookie = $config['cookie'] ?? '';
$apikey = $config['apikey'] ?? '';

if (empty($apikey)) {
    echo KUNING . "[*] API Key ga ada. Masukin API Key: " . RESET;
    $apikey = get_apikey_from_user();
    $config['apikey'] = $apikey;
    save_config($config);
}

while (true) {
    if (empty($cookie)) {
        echo KUNING . "[*] Cookie kosong. Masukin cookie baru:" . RESET . "\n";
        $cookie = get_cookie_from_user();
        $config['cookie'] = $cookie;
        save_config($config);
    }
    
    $jar = new CookieJar();
    $jar->fromString($cookie);
    
    echo CYAN . "[*] Validasi cookie..." . RESET . "\n";
    
    if (!$jar->get('captcha')) {
        echo KUNING . "[!] Cookie 'captcha' tidak ditemukan. Ini penting!" . RESET . "\n";
        echo KUNING . "[*] Pastikan cookie dari browser sudah include 'captcha=adslab'" . RESET . "\n";
    }
    
    if (is_logged_in($jar)) {
        echo HIJAU . "[+] Cookie valid." . RESET . "\n";
        $updated_cookie = $jar->toString();
        if ($updated_cookie !== $config['cookie']) {
            $config['cookie'] = $updated_cookie;
            save_config($config);
            echo CYAN . "[*] Cookie di-update dari server." . RESET . "\n";
        }
        break;
    } else {
        echo MERAH . "[!] Cookie ga valid atau expired." . RESET . "\n";
        echo KUNING . "[*] Cek file debug_*.html untuk melihat respon server." . RESET . "\n";
        echo KUNING . "[*] Masukin cookie baru:" . RESET . "\n";
        $cookie = get_cookie_from_user();
        $config['cookie'] = $cookie;
        save_config($config);
        continue;
    }
}

$coins = get_coins($jar);
if (empty($coins)) {
    echo MERAH . "[!] Gagal ambil daftar coin. Cek cookie." . RESET . "\n";
    exit(1);
}

echo "\n" . KUNING . "💰 Daftar coin yang tersedia:" . RESET . "\n";
$i = 1;
foreach ($coins as $coin) {
    printf("%s(%2d) %s%-6s%s", PUTIH, $i, HIJAU, strtoupper($coin), RESET);
    if ($i % 4 == 0 || $i == count($coins)) echo "\n";
    $i++;
}
echo "\n" . CYAN . "🎯 Pilih nomor coin: " . RESET;
$choice = trim(fgets(STDIN));
$idx = (int)$choice - 1;
$coin = isset($coins[$idx]) ? $coins[$idx] : $coins[0];
echo HIJAU . "✅ Coin dipilih: " . strtoupper($coin) . RESET . "\n";

$count = 0;
$failures = 0;   // <-- counter gagal berturut-turut

while (true) {
    $count++;
    echo "\n" . CYAN . "┌─[ ROUND $count ]" . RESET . "\n";
    $result = claim_faucet($jar, $coin, $apikey);
    
    $current_cookie = $jar->toString();
    if ($current_cookie !== $config['cookie']) {
        $config['cookie'] = $current_cookie;
        save_config($config);
        echo CYAN . "[*] Cookie state disimpan." . RESET . "\n";
    }
    
    if ($result === 'EXPIRED') {
        // Session expired, minta cookie baru, reset failure
        $failures = 0;
        echo KUNING . "[*] Cookie expired. Masukin cookie baru:" . RESET . "\n";
        $cookie = get_cookie_from_user();
        $config['cookie'] = $cookie;
        save_config($config);
        
        $jar = new CookieJar();
        $jar->fromString($cookie);
        
        if (!is_logged_in($jar)) {
            echo MERAH . "[!] Cookie masih ga valid. Coba lagi." . RESET . "\n";
            continue;
        }
        echo HIJAU . "[+] Cookie baru valid." . RESET . "\n";
        $config['cookie'] = $jar->toString();
        save_config($config);
        continue;
    }
    
    if ($result === true) {
        // Sukses, reset failure
        $failures = 0;
        echo HIJAU . "⏳ Tunggu buat claim berikutnya..." . RESET . "\n";
        timer(CLAIM_INTERVAL, "🔄 Claim berikutnya dalam");
    } else {
        // Gagal (cooldown, error, dll)
        $failures++;
        echo KUNING . "[!] Gagal ke-$failures dari " . MAX_FAILURES . " (maks)." . RESET . "\n";
        if ($failures >= MAX_FAILURES) {
            echo MERAH . "[!] Bot berhenti karena sudah 2x gagal berturut-turut." . RESET . "\n";
            exit(0);
        }
        echo KUNING . "🔄 Coba lagi dalam " . RETRY_INTERVAL . " detik..." . RESET . "\n";
        timer(RETRY_INTERVAL, "🔄 Coba lagi dalam");
    }
}

File: altcryp.py
#!/usr/bin/env python3
import requests
import time
import re
import json
import sys
import os
from bs4 import BeautifulSoup

# ========== WARNA ANSI ==========
C = {
    "hitam": "\033[0;30m",
    "merah": "\033[0;31m",
    "hijau": "\033[0;32m",
    "kuning": "\033[0;33m",
    "biru": "\033[0;34m",
    "cyan": "\033[0;36m",
    "putih": "\033[0;37m",
    "bold": "\033[1m",
    "reset": "\033[0m",
    "dim": "\033[2m",
}

HEADER = f"""
{C['cyan']}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{C['reset']}
{C['bold']}{C['kuning']}              🚀 ALTCRYP AUTO BOT{C['reset']}
{C['cyan']}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{C['reset']}
"""

SEPARATOR = f"{C['cyan']}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{C['reset']}"

def print_box(title, lines, color=C['hijau']):
    print(f"{C['bold']}{color}╭─[ {title} ]{C['reset']}")
    for line in lines:
        print(f"{color}│{C['reset']} {line}")
    print(f"{color}╰────────────────────────────────────────────────────{C['reset']}")

def timer(seconds, prefix="[!] Please wait"):
    wait_time = int(seconds)
    frames = ['⣾', '⣽', '⣻', '⢿', '⡿', '⣟', '⣯', '⣷']
    frame_count = len(frames)
    current_frame = 0
    while wait_time > 0:
        start = time.time()
        while (time.time() - start) < 1:
            hours = wait_time // 3600
            minutes = (wait_time % 3600) // 60
            secs = wait_time % 60
            time_str = f"{hours:02d}:{minutes:02d}:{secs:02d}"
            spinner = frames[current_frame]
            print(f"{C['putih']}{prefix} {C['hijau']}{time_str} {C['putih']}{spinner}\r", end="")
            time.sleep(0.1)
            current_frame = (current_frame + 1) % frame_count
            if (time.time() - start) >= 1:
                break
        wait_time -= 1
    print(" " * 50 + "\r", end="")

def clear():
    os.system('clear' if os.name == 'posix' else 'cls')

# ========== KONFIGURASI ==========
BASE_URL = "https://altcryp.com"
SOLVER_BASE = "https://bypassallshortlinks.space"
CONFIG_FILE = "config_altcryp.json"
CLAIM_INTERVAL = 14  # <-- diubah jadi 14 detik

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Mobile Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
    "Accept-Language": "id-ID",
    "Accept-Encoding": "gzip, deflate, br",
    "Sec-Ch-Ua": '"Chromium";v="127", "Not)A;Brand";v="99"',
    "Sec-Ch-Ua-Mobile": "?1",
    "Sec-Ch-Ua-Platform": "Android",
}

def get_config():
    if os.path.exists(CONFIG_FILE):
        with open(CONFIG_FILE, 'r') as f:
            return json.load(f)
    else:
        print(f"{C['putih']}Cookie (dari browser): {C['kuning']}", end="")
        cookie = input().strip()
        print(f"{C['putih']}API Key bypassallshortlinks: {C['kuning']}", end="")
        api_key = input().strip()
        config = {"cookie": cookie, "apikey": api_key}
        with open(CONFIG_FILE, 'w') as f:
            json.dump(config, f, indent=4)
        print(f"{C['hijau']}Config disimpan ke {CONFIG_FILE}{C['reset']}")
        time.sleep(1)
        return config

def solve_turnstile(sitekey, pageurl, apikey, show_progress=False):
    submit = requests.get(f"{SOLVER_BASE}/in.php", params={
        "key": apikey,
        "method": "turnstile",
        "sitekey": sitekey,
        "pageurl": pageurl
    }, timeout=30)
    if not submit.text.startswith("OK|"):
        print(f"{C['merah']}[!] Gagal submit Turnstile: {submit.text}{C['reset']}")
        return None
    task_id = submit.text.split("|")[-1]
    if show_progress:
        print(f"{C['kuning']}[+] Task ID    : {task_id}{C['reset']}")
        print(f"{C['cyan']}[*] Menunggu Turnstile...{C['reset']}")
    for i in range(45):
        time.sleep(2)
        poll = requests.get(f"{SOLVER_BASE}/res.php", params={"id": task_id, "key": apikey}, timeout=30)
        if poll.text.startswith("OK|"):
            token = poll.text.split("|")[-1]
            if show_progress:
                print(f"{C['hijau']}[+] Turnstile  : SOLVED ✓{C['reset']}")
            return token
        if show_progress and i % 5 == 0:
            print(f"{C['dim']}[*] Polling ke-{i+1}/45...{C['reset']}")
    print(f"{C['merah']}[!] Turnstile timeout.{C['reset']}")
    return None

def get_coins(session):
    resp = session.get(BASE_URL, headers=HEADERS, timeout=30)
    if resp.status_code != 200:
        return []
    soup = BeautifulSoup(resp.text, "html.parser")
    links = soup.find_all("a", href=True)
    coins = []
    for link in links:
        href = link.get("href")
        if href and "/faucet/currency/" in href:
            coin = href.split("/faucet/currency/")[-1]
            if coin and coin not in coins:
                coins.append(coin)
    return coins

def get_csrf_and_token(session, coin):
    url = f"{BASE_URL}/faucet/currency/{coin}"
    resp = session.get(url, headers=HEADERS, timeout=30)
    if resp.status_code != 200:
        return None, None
    soup = BeautifulSoup(resp.text, "html.parser")
    csrf_input = soup.find("input", {"name": "csrf_token_name"})
    csrf = csrf_input.get("value") if csrf_input else None
    token_input = soup.find("input", {"name": "token"})
    token = token_input.get("value") if token_input else None
    if not csrf:
        match = re.search(r'name="csrf_token_name"\s+value="([^"]+)"', resp.text)
        if match:
            csrf = match.group(1)
    if not token:
        match = re.search(r'name="token"\s+value="([^"]+)"', resp.text)
        if match:
            token = match.group(1)
    return csrf, token

def claim_faucet(session, coin, apikey):
    print(f"{C['cyan']}[*] Claim      : PROCESSING... (coin: {coin.upper()}){C['reset']}")
    csrf, token = get_csrf_and_token(session, coin)
    if not csrf or not token:
        print(f"{C['merah']}[!] Gagal ambil CSRF/Token.{C['reset']}")
        return False
    print(f"{C['kuning']}[+] CSRF Token : {csrf}{C['reset']}")
    print(f"{C['kuning']}[+] Token      : {token}{C['reset']}")

    sitekey = "0x4AAAAAAAHPLPJjjJUpAitl"
    pageurl = f"{BASE_URL}/faucet/currency/{coin}"
    turnstile_token = solve_turnstile(sitekey, pageurl, apikey, show_progress=True)
    if not turnstile_token:
        return False

    data = {
        "username_fake_field": "",
        "csrf_token_name": csrf,
        "token": token,
        "captcha": "turnstile",
        "cf-turnstile-response": turnstile_token
    }

    url_verify = f"{BASE_URL}/faucet/verify/{coin}"
    resp = session.post(url_verify, data=data, headers={
        **HEADERS,
        "Content-Type": "application/x-www-form-urlencoded",
        "Origin": BASE_URL,
        "Referer": f"{BASE_URL}/faucet/currency/{coin}",
    }, timeout=30)

    if "Good job" in resp.text or "success" in resp.text.lower():
        print(f"{C['hijau']}[+] Success sent to faucetpay ✓{C['reset']}")  # <-- ubah reward message
        return True
    elif "already claimed" in resp.text.lower() or "wait" in resp.text.lower():
        print(f"{C['kuning']}[!] Claim      : COOLDOWN ⏳{C['reset']}")
        return False
    else:
        print(f"{C['merah']}[?] Claim tidak jelas.{C['reset']}")
        with open("claim_debug_altcryp.html", "w", encoding="utf-8") as f:
            f.write(resp.text)
        return False

def main():
    clear()
    print(HEADER)
    print(SEPARATOR)
    print()

    config = get_config()
    cookie = config.get('cookie', '')
    apikey = config.get('apikey', '')

    if not cookie or not apikey:
        print(f"{C['merah']}[!] Cookie atau API Key kosong.{C['reset']}")
        return

    session = requests.Session()
    session.headers.update({"Cookie": cookie})

    coins = get_coins(session)
    if not coins:
        print(f"{C['merah']}[!] Gagal ambil daftar coin. Cek cookie.{C['reset']}")
        return

    print(f"{C['kuning']}💰 Daftar coin yang tersedia:{C['reset']}")
    for i, coin in enumerate(coins, 1):
        num = str(i).rjust(2)
        coin_upper = coin.upper().ljust(6)
        print(f"{C['putih']}({num}) {C['hijau']}{coin_upper}{C['reset']}", end=" ")
        if i % 4 == 0 or i == len(coins):
            print()
    print()
    choice = input(f"{C['cyan']}🎯 Pilih nomor coin: {C['reset']}").strip()
    try:
        idx = int(choice) - 1
        coin = coins[idx]
    except:
        print(f"{C['merah']}[!] Pilihan tidak valid. Default ke coin pertama.{C['reset']}")
        coin = coins[0]
    print(f"{C['hijau']}✅ Coin dipilih: {coin.upper()}{C['reset']}")

    count = 0
    while True:
        count += 1
        print()
        print_box(f"🎯 ROUND {count}", [
            f"{C['cyan']}🚀 Claim       : PROCESSING... (coin: {coin.upper()}){C['reset']}"
        ], C['cyan'])
        success = claim_faucet(session, coin, apikey)
        if success:
            print(f"{C['hijau']}⏳ Waiting for next claim...{C['reset']}")
            timer(CLAIM_INTERVAL, prefix="🔄 Next claim in")  # <-- pakai CLAIM_INTERVAL = 14 detik
        else:
            print(f"{C['kuning']}🔄 Retry in 60 seconds...{C['reset']}")
            timer(60, prefix="🔄 Retry in")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print(f"\n{C['merah']}[!] Bot dihentikan oleh user.{C['reset']}")
        sys.exit(0)
